gotham/src/router/tree/node.rs

```diff
@@ -385,11 +393,22 @@ impl NodeBuilder {
         children.shrink_to_fit();
         self.routes.shrink_to_fit();
 
+        let mut allow_header_method_list: Vec<Method> = self.routes
+            .iter()
+            .flat_map(|r| r.allow_header_method_list())
+            .collect();
+
+        allow_header_method_list
+            .sort_by(|a, b| AsRef::<str>::as_ref(a).cmp(AsRef::<str>::as_ref(b)));
+
+        allow_header_method_list.dedup();
```

Comment:
I was re-reading https://tools.ietf.org/html/7231#section-7.4.1

> The actual set of allowed methods is defined by the origin server at
   the time of each request.  An origin server MUST generate an Allow
   field in a 405 (Method Not Allowed) response and MAY do so in any
   other response.  An empty Allow field value indicates that the
   resource allows no methods, which might occur in a 405 response if
   the resource has been temporarily disabled by configuration.

Which raised 2 thoughts:

1. In the case where Gotham is matching requests but has not had any methods explicitly set, should we be returning all HTTP verbs instead of the empty set
2. Allow is a potential header for non 405 responses as well, probably worth altering associated comments to remove implication that this is only applicable for 405.

Reply:
> In the case where Gotham is matching requests but has not had any methods explicitly set, should we be returning all HTTP verbs instead of the empty set

If that's the case, the `RouteMatcher` shouldn't be returning a `405 Method Not Allowed`, and the `Allow` header code won't be triggered. If a `RouteMatcher` permits all methods and then returns a 405, that's a whole 'nother problem 😁 

> Allow is a potential header for non 405 responses as well, probably worth altering associated comments to remove implication that this is only applicable for 405.

Perhaps, but that's the only way it's used *currently*, and it's populated with only that usage in mind. Other use cases would probably be incorrect due to the first point, and the fact that it's no longer only triggered by a 405 response.

Reply:
> > In the case where Gotham is matching requests but has not had any methods explicitly set, should we be returning all HTTP verbs instead of the empty set
>
> If that's the case, the `RouteMatcher` shouldn't be returning a `405 Method Not Allowed`, and the `Allow` header code won't be triggered. If a `RouteMatcher` permits all methods and then returns a 405, that's a whole 'nother problem 😁 

To look at this another way:

```
GET / HTTP/1.1
Host: x.example.com
```

```
HTTP/1.1 405 Method Not Allowed
Allow: GET, HEAD, POST, PATCH, ... [all methods]
```

Doesn't make sense.

Reply:
I may have been less than clear in my description. The intent was to take this back to being generic, as it seems to me that it actually is, aside from populating all methods if `allow_header_method_list` is empty.

Right now the router check is as follows:

```
if let StatusCode::MethodNotAllowed = status ...
```

Which should only be the response status (custom implementations doing stupid things aside) if `MethodOnlyRouteMatcher` decides the request is not satisfactory. To do that it follows that `allow_header_method_list` must have been populated during tree construction, hence the 405 header would be correctly populated and your example above would not occur.

For folks that wanted to populate an `Allow` header for some other purpose `Allow(leaf.allow_header_method_list())` would then be just as applicable. Albeit accessing that API from something other than the router would be hard(impossible? different discussion...) for now.

I'm not bound to this, I just think there is an opportunity here to make this code more widely applicable (but likely only ever used by 405 in reality).

Reply:
> I'm not bound to this, I just think there is an opportunity here to make this code more widely applicable (but likely only ever used by 405 in reality).

I think that altering it to be more widely applicable is a minefield of possibilities. For example:

```
/a [MethodOnlyRouteMatcher - GET, POST]
/a [AnyRouteMatcher]
```

The `Allow` list in that case would only be populated with `[GET, POST]`. The fact that this list is strictly incorrect is irrelevant to the implementation because the HTTP method is always matched by the second route, so the `405` never fires.

The logic makes lists that are applicable for creating `Allow` headers for `405` response, and not always applicable for other purposes. Rather than try to engineer it to be correct in all cases, I just documented it as only being for `405` and left it as-is.

Making it applicable for all cases doesn't seem doable without generating the list on a per-request basis. That's the implementation I started with (the `Result<(), StatusCode>` became a `Result<(), NonMatch>` with appropriate metadata being propagated in the `NonMatch` enum), but dropped back to this when the change proved to be quite far-reaching.

If we want the ability to extend the usage of the `Allow` header later, I think we'll need to go back to this version of it. The question is whether we need it.

It will make the implementation more complex, but also more internally consistent.

Maybe it's a good idea anyway.